Added queuing of controller profiles.

LemmusLemmus opened the pull request:
Profiles whose names end with “_№” (for example “Nunchuk_Xinput_IR_1.wup”) will be automatically switched between depending on the currently connected devices. The number should correspond with the selected rumble controller of the profile. This functionality is currently only available for Nintendo controllers.

This is useful when you have several controllers whose profiles should differ depending on the already connected controllers. If, for example, four Wii Remotes are connected to WiinUPro they will be assigned the profiles in their corresponding order (1, 2, 3, 4) in accordance with the order they are connected to WiinUPro, irrespective of which particular Wiimote is already connected.

Instead of having to bind Wiimote A, B and C to Player1, Player2 and Player3 respectively, each Wiimote can be bound to all respective player numbers. If one, for example, decides to use only two Wiimotes it no longer matters whether these two are Wiimotes A & B or Wiimotes A & C (Connecting A & C would previously assign them to profiles corresponding to Player1 and Player3, in which case an additional unused XInput device would be added).

This system is relatively rudimentary and should not interfere with anything. If an overload profile is desired the name of the profile can end with “_5”.

LemmusLemmus commented on WiinUPro/Controls/NintyControl.xaml.cs:
```diff
@@ -394,6 +394,32 @@ public void LoadProfile(string fileName)
+                {
+                    int n = (int)Char.GetNumericValue(fileName[fileName.Length - 5]);
+                    bool changeProfile = false;
+                    if (n == 1 && ScpDirector.Access.IsConnected(ScpDirector.XInput_Device.Device_A))
+                    {
+                        changeProfile = true;
+                    }
+                    else if (n == 2 && ScpDirector.Access.IsConnected(ScpDirector.XInput_Device.Device_B))
+                    {
+                        changeProfile = true;
+                    }
+                    else if (n == 3 && ScpDirector.Access.IsConnected(ScpDirector.XInput_Device.Device_C))
+                    {
+                        changeProfile = true;
+                    }
+                    else if (n == 4 && ScpDirector.Access.IsConnected(ScpDirector.XInput_Device.Device_D))
+                    {
+                        changeProfile = true;
+                    }
```
A way to iterate through the connected devices would simplify this code. Such iteration would also simplify the code relating to the connecting of these devices in MainWindow.xaml.cs.

LemmusLemmus commented on WiinUPro/Controls/NintyControl.xaml.cs:
```diff
@@ -394,6 +394,32 @@ public void LoadProfile(string fileName)
+                    {
+                        changeProfile = true;
+                    }
+                    if (changeProfile && !_rumbleSubscriptions[n - 1])
```
If it were not for the extension specific profiles merely checking whether or not a profile has already been loaded for the controller would suffice, since the check would only have to be performed once, when the first profile is loaded. The existence of extension specific profiles necessitates that the rumble controller corresponds with the number of the profile. This could probably be avoided somehow, but I personally have no use case where I would assign an extension's rumble controller to a different one than the its corresponding XInput device.

LemmusLemmus commented on WiinUPro/Controls/NintyControl.xaml.cs:
```diff
@@ -394,6 +394,32 @@ public void LoadProfile(string fileName)
+                    }
+                    if (changeProfile && !_rumbleSubscriptions[n - 1])
+                    {
+                        LoadProfile(fileName.Substring(0, fileName.Length - 5) + (char)(49 + n) + ".wup");
```
(char)48 = 0, (char)49 = 1 etc.

KeyPuncher replied:
Cleaver, but easier to read would be taking that last split part mentioned above then using a Replace with the part string  and then a string.Format or what's called an interpolated string `$"_{n}.wup"`

KeyPuncher commented:
This feature should be a setting that you can enable and disable.

KeyPuncher commented on WiinUPro/Controls/NintyControl.xaml.cs:
```diff
@@ -394,6 +394,32 @@ public void LoadProfile(string fileName)
 
             if (loadedProfile != null)
             {
+                if (fileName.Length > 6 & fileName[fileName.Length - 6] == '_')
```
Try using the string Split function instead.

1. Split on the underscore character
2. Take the last part of the split array and remove the file extension
3. Use TryParse to check if it results in a number between 1 and 4

LemmusLemmus replied:
Like so?
```C#
string[] subStrings = fileName.Split('_');
if(Int32.TryParse(subStrings[subStrings.Length - 1].Remove(subStrings[subStrings.Length - 1].Length - 4), out int n) && n > 0 && n < 5)
          {
              // Rest of code goes here
          }
```

One might still want to check for the length of the filename of the profile, as "1.wup" for example would otherwise, arguably incorrectly, be recognized as one among other queueable profiles.

KeyPuncher commented on WiinUPro/Controls/NintyControl.xaml.cs:
```diff
@@ -394,6 +394,32 @@ public void LoadProfile(string fileName)
+                {
+                    int n = (int)Char.GetNumericValue(fileName[fileName.Length - 5]);
+                    bool changeProfile = false;
+                    if (n == 1 && ScpDirector.Access.IsConnected(ScpDirector.XInput_Device.Device_A))
```
You can instead check if n is between (int)Deivce_A and (int)Device_D

LemmusLemmus replied:
I'm not sure if I follow. Do you merely wish for 1, 2, 3 & 4 to be replaced with (int)Device_A, (int)Device_B etc. or do you suggest further restructuring of lines 401-416? If so, in what manner?

If you meant checking whether or not the value of n is between 1 ((int)Deivce_A) and 4 ((int)Device_D), then I'm unsure as to what to do with this information, as it doesn't tell me much about whether or not a profile change is due.

On a side note, since all the if-statements have the same result it would be possible to combine all the checks into a single if-statement, although said if-statement would become quite long.

One may alternatively rewrite this section of code as to not rely on recursion.

KeyPuncher replied:
Yeah, you can actually reduce this down quite a bit like so

```suggestion
                string[] fileNameParts = fileName.Split('_');
                string fileNameEnding = fileNameParts[fileNameParts.Length - 1].Replace(".wup", "");
                
                if (int.TryParse(fileNameEnding, out int profileNum)
                    && profileNum >= (int)ScpDirector.XInput_Device.Device_A
                    && profileNum <= (int)ScpDirector.XInput_Device.Device_D
                    && ScpDirector.Access.IsConnected((ScpDirector.XInput_Device)profileNum)
                {
                    // load a different profile
                }
```

Also, what's the behavior if all four controllers are connected? Looks for profile_5.wup and fails if it doesn't exist?

LemmusLemmus replied:
Do you prefer using replace or removing the last 4 characters to get rid of the ".wup"?
```C#
ScpDirector.Access.IsConnected((ScpDirector.XInput_Device)profileNum)
```
I didn't know that was possible! Your code snippet above is the behavior I have been looking for which would allow for code simplification here and in MainWindow.xaml.cs.


Due to the possibility of extensions changing the profile the additional check below is also necessary to add to your suggested code, other than that it seems perfect.
```C#
&& !_rumbleSubscriptions[profileNum - 1]
```
_The code above is only one potential way to go about handling extension specific profiles._

KeyPuncher replied:
Remove is good too. I actually forgot about it because I use Replace so frequently.

Ah yes, I saw that it should be added when I was reading it but forgot to add it to the actual example.

KeyPuncher commented on WiinUPro/Windows/DevicePrefsWindow.xaml:
```diff
@@ -32,5 +32,18 @@
         <Label x:Name="labelExtProfile" Content="Extension Specific Profile:" HorizontalAlignment="Left" Margin="10,114,0,0" VerticalAlignment="Top" FontWeight="Bold" Foreground="{StaticResource TextHeader}"/>
         <TextBox x:Name="extProfile" Height="23" Margin="10,167,10,0" TextWrapping="Wrap" Text="" VerticalAlignment="Top" Foreground="#FF444444"/>
         <Button x:Name="btnExtProfile" Content=". . ." HorizontalAlignment="Right" Margin="0,144,10,0" VerticalAlignment="Top" Width="32" Style="{StaticResource BasicButton}" Click="btnExtProfile_Click"/>
+        <TextBlock Text="№ Specific Profile" HorizontalAlignment="Left" Margin="119,68,0,0" VerticalAlignment="Top" Foreground="{DynamicResource BackgroundMain}" ToolTip="Profiles whose names end with &quot;_№&quot; (for example &quot;Nunchuk_Xinput_IR_1.wup&quot;) will be automatically switched between depending on the currently connected devices.&#x0a;The number should correspond with the selec ...
```
Can we split this line up, it's pretty long. I haven't been super good about that but I'm gonna eventually try and tackle optimization opportunities as I run into them.

LemmusLemmus commented:
Shall I update my fork to the current version of WiinUPro before I make any changes to MainWindow.xaml.cs in my fork?

KeyPuncher commented:
Yeah, you'll want to update it first. I don't believe there will be conflicts so you shouldn't have to do any complicated merging.

LemmusLemmus commented on WiinUPro/Windows/MainWindow.xaml:
```diff
@@ -72,11 +72,13 @@
+                            <CheckBox x:Name="settingToshibaMode" Content="Use Toshiba Mode" HorizontalAlignment="Left" Margin="10,90,0,0" VerticalAlignment="Top" Style="{StaticResource CheckBoxStyle}" Click="settingToshibaMode_Checked"/>
+                            <CheckBox x:Name="settingSuppressLostConn" Content="Hide Connection Errors" HorizontalAlignment="Left" Margin="10,110,0,0" VerticalAlignment="Top" Style="{StaticResource CheckBoxStyle}" Click="settingSuppressLostConn_Checked"/>
+                            <CheckBox x:Name="settingAutoAddXInputDevices" Content="Auto Add XInput Devices" HorizontalAlignment="Left" Margin="10,130,0,0" VerticalAlignment="Top" Style="{StaticResource CheckBoxStyle}" Click="settingAutoAddXInputDevices_Checked" ToolTip="Automatically connect emulated XInput Devices based on device rumble configurations."/>
+                            <CheckBox x:Name="settingMinimizeOnClose" Content="Minimize To System Tray" HorizontalAlignment="Left" Margin="10,50,0,0" VerticalAlignment="Top" Style="{StaticResource CheckBoxStyle}" Click="settingMinimizeOnClose_Checked" ToolTip="Closing the application will instead minimize it to the system tray."/>
```
I did some slight reordering of the buttons, the "Minimize To System Tray" button is now right under the "Start Minimized" button.

LemmusLemmus commented on WiinUPro/Windows/MainWindow.xaml.cs:
```diff
@@ -375,12 +375,17 @@ private void settingSuppressLostConn_Checked(object sender, RoutedEventArgs e)
         private void settingAutoAddXInputDevices_Checked(object sender, RoutedEventArgs e)
         {
             AppPrefs.Instance.autoAddXInputDevices = settingAutoAddXInputDevices.IsChecked ?? false;
+            settingProfileQueuing.IsEnabled = settingAutoAddXInputDevices.IsChecked ?? false;
```
The option of Profile Queueing is now only enabled when autoAddXInputDevices is, since the Profile Queueing will not work particularly well without it.

KeyPuncher replied:
That makes sense. Although I think the way this is setup is that if you check or uncheck the auto add XInput devices then it also checks or unchecks auto profile queuing. So if profile queuing is off and then auto add XInput gets checked, profile queuing automatically gets turned on which probably shouldn't be the expected behvaior.

Here I think the only other operation we want to do is check if auto add xinput is being turned off and if it is make sure profile queuing also gets turned off.

KeyPuncher replied:
Oh my bad, this is right. you're changing IsEnabled, all good!

LemmusLemmus commented:
Everything should have been merged now and all of your suggestions should have hopefully been properly addressed!

On a different note, the MainWindow.xaml.cs code I personally use differs slightly from the one on GitHub as I've added a function which automatically "suspends" WiinUPro (actually just disconnecting all devices, and reconnecting them again when it is "unsuspended") when the Dolphin emulator is running, as such I rewrote my changes in the updated file (hopefully I did not miss any). Would a feature which temporarily "suspends" WiinUPro when a chosen program is running be desirable to add as an optional feature in the future, or is the use case too niche?

LemmusLemmus commented on WiinUPro/Windows/MainWindow.xaml:
```diff
@@ -72,11 +72,13 @@
+                            <CheckBox x:Name="settingAutoAddXInputDevices" Content="Auto Add XInput Devices" HorizontalAlignment="Left" Margin="10,130,0,0" VerticalAlignment="Top" Style="{StaticResource CheckBoxStyle}" Click="settingAutoAddXInputDevices_Checked" ToolTip="Automatically connect emulated XInput Devices based on device rumble configurations."/>
+                            <CheckBox x:Name="settingMinimizeOnClose" Content="Minimize To System Tray" HorizontalAlignment="Left" Margin="10,50,0,0" VerticalAlignment="Top" Style="{StaticResource CheckBoxStyle}" Click="settingMinimizeOnClose_Checked" ToolTip="Closing the application will instead minimize it to the system tray."/>
+                            <CheckBox x:Name="settingProfileQueuing" Content="Profile Queuing" HorizontalAlignment="Left" Margin="10,150,0,0" VerticalAlignment="Top" Style="{StaticResource CheckBoxStyle}" Click="settingProfileQueuing_Checked" IsEnabled="False"
+                                      ToolTip="Profiles whose names end with &quot;_№&quot; (for example &quot;Nunchuk_Xinput_IR_1.wup&quot;) will be automatically switched between depending on the currently connected devices.&#x0a;The number should correspond with the selected rumble controller of the profile. This functionality is only available for Nintendo controllers."/>
```
The tooltip is now on a separate line, although it could use some reformulation.
